File: ohms/answer.py
#  This Source Code Form is subject to the terms of the Mozilla Public
#  License, v. 2.0. If a copy of the MPL was not distributed with this
#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
#  Copyright (c) 2012-2013, Dennis Sun <dlsun@stanford.edu>

import random
import re

class Answer(object):

    expected_type = unicode

    def cast(self,response):
        try:
             return self.expected_type(response)
        except:
             if self.expected_type==int: return -1
             else: return self.expected_type("inf")        

class String(Answer):

    def cast(self,response):
        return super(String,self).cast(response.rstrip())


class NumberAnswer(Answer):

    def cast(self,response):
        response = response.replace(" ","")
        out = re.match("[^\d.-]*(-?[\d.]+)",response)
        if out:
            return super(NumberAnswer,self).cast(out.groups()[0])
        else:
            return super(NumberAnswer,self).cast("")

class Integer(NumberAnswer):
    expected_type = int
    

class Float(NumberAnswer):
    expected_type = float
        

class TrueFalseAnswer(Answer):
    expected_type = unicode

class Text(Answer):
    expected_type = unicode

class MultipleChoiceAnswer(Answer):
    expected_type = int
    compact = False

    def __init__(self,choices):
        self.choices = choices

File: WWW/js/answer.js
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  License, v. 2.0. If a copy of the MPL was not distributed with this
 *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 *  Copyright (c) 2012-2013, Dennis Sun <dlsun@stanford.edu>
 */

var OHMS = (function(OHMS) {

    var Answer = function (question) {
	this.element = this.element || null;
	this.question = question;
	this.create_element();
	this.element.data("answer",this);
	this.bind_events();
	if (this.solution !== undefined)
	    this.show_solution();
    }

    Answer.prototype.create_element = function () {
	this.element = $("<input type='text' name='ans' class='input-small' disabled='true'>");
    }

    Answer.prototype.bind_events = function () {
    }

    Answer.prototype.show_solution = function () {
    }

    Answer.prototype.get_value = function () {
	return this.element.val();
    }

    Answer.prototype.set_value = function (value) {
	this.element.val(value);
    }

    Answer.prototype.unlock_answer = function () {
	this.element.removeAttr("disabled");
    }

    Answer.prototype.lock_answer = function () {
	this.element.attr("disabled","disabled");
    }

    OHMS.Answer = Answer;

    return OHMS;

}(OHMS))